internal/eval/eval.go

```diff
@@ -675,6 +675,162 @@ func evalJSONSET(args []string, store *dstore.Store) []byte {
+// in the JSON object saved at key in arguments.
+// Args must contain atleast a key, path and value.
+// If the key does not exist or is expired, it returns response.RespNIL.
+// If the object at given path is not an array, it returns response.RespNIL.
+// Returns the new length of the array at path.
+func evalJSONARRAPPEND(args []string, store *dstore.Store) []byte {
+    if len(args) < 3 {
+        return diceerrors.NewErrArity("JSON.ARRAPPEND")
+    }
+
+    var key string = args[0]
+    var path string = args[1]
+
+    obj := store.Get(key)
+    if obj == nil {
+        return diceerrors.NewErrWithMessage("could not perform this operation on a key that doesn't exist")
+    }
+
+    errWithMessage := dstore.AssertTypeAndEncoding(obj.TypeEncoding, dstore.ObjTypeJSON, dstore.ObjEncodingJSON)
+    if errWithMessage != nil {
+        return errWithMessage
+    }
+
+    jsonData := obj.Value
+    _, err := sonic.Marshal(jsonData)
+    if err != nil {
+        return diceerrors.NewErrWithMessage("Existing key has wrong Dice type")
+    }
+
+    expr, err := jp.ParseString(path)
+    if err != nil {
+        return diceerrors.NewErrWithMessage("invalid JSONPath")
+    }
+
+    resultsArray := make([]interface{}, 0)
+    results := expr.Get(jsonData)
+    if(len(results) == 0){
+        return clientio.Encode(resultsArray, false)
+    }
+
+    if path == defaultRootPath {
+        pathArrData, ok := jsonData.([]interface{})
+        if(!ok) {
+            return clientio.RespNIL
+        }
+        for _, arg := range args[2:] {
+            var value interface{}
+            var conversionErr error
+
+            // Determine type of element in the existing array (if any) to match type
+            if len(pathArrData) > 0 {
+                switch utils.GetJSONFieldType(pathArrData[0]) {
+                case utils.StringType:
+                    value = arg
+                case utils.NumberType:
+                    value, conversionErr = strconv.ParseFloat(arg, 64)
+                case utils.IntegerType:
+                    var intVal int64
+                    intVal, conversionErr = strconv.ParseInt(arg, 10, 64)
+                    value = intVal
+                case utils.ArrayType:
+                    var arrObject []interface{}
+                    conversionErr = sonic.Unmarshal([]byte(arg), &arrObject)
+                    if conversionErr == nil {
+                        value = arrObject
+                    }
+                case utils.ObjectType:
+                    var jsonObject map[string]interface{}
+                    conversionErr = sonic.Unmarshal([]byte(arg), &jsonObject)
+                    if conversionErr == nil {
+                        value = jsonObject
+                    }
+                default:
+                    return diceerrors.NewErrWithMessage("failed to convert existing value in array to a valid json field type")
```

Comment:
Redis does not throw an error in case of incompatible types existing in a json array.